### PR TITLE
[SPARK-44694][PYTHON][CONNECT] Refactor active sessions and expose them as an API

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to (mostly) refactor all the internal workarounds to get the active session correctly.

There are few things to note:

- _PySpark without Spark Connect does not already support the hierarchy of active sessions_. With pinned thread mode (enabled by default), PySpark does map each Python thread to JVM thread, but the thread creation happens within gateway server, that does not respect the thread hierarchy. Therefore, this PR follows the exactly same behaviour.
  - New thread will not have an active thread by default.
  - Other behaviours are same as PySpark without Connect, see also https://github.com/apache/spark/pull/42367
- Since I am here, I piggiyback few documentation changes. We missed document `SparkSession.readStream`, `SparkSession.streams`, `SparkSession.udtf`, `SparkSession.conf` and `SparkSession.version` in Spark Connect.
- The changes here are mostly refactoring that reuses existing unittests while I expose two methods:
  - `SparkSession.getActiveSession` (only for Spark Connect)
  - `SparkSession.active` (for both in PySpark)

### Why are the changes needed?

For Spark Connect users to be able to play with active and default sessions in Python.

### Does this PR introduce _any_ user-facing change?

Yes, it adds new API:
  - `SparkSession.getActiveSession` (only for Spark Connect)
  - `SparkSession.active` (for both in PySpark)

### How was this patch tested?

Existing unittests should cover all.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @hvanhovell @ueshin @allisonwang-db @rangadi @WeichenXu123 @xinrong-meng @zhengruifeng @itholic. I happened to touch here and there - would be great if you guys find some time to take a look.

File: python/pyspark/errors/error_classes.py [hunk 1]
```diff
@@ -622,6 +622,11 @@
       "No active Spark session found. Please create a new Spark session before running the code."
     ]
   },
+  "NO_ACTIVE_OR_DEFAULT_SESSION" : {
```
Comment (itholic) [inline python/pyspark/errors/error_classes.py:625]: We already have `NO_ACTIVE_SESSION`. Maybe we can update the existing error class?

**Review by itholic — APPROVED** [review]
Looks good otherwise

File: python/pyspark/errors/error_classes.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/errors/error_classes.py:625]: How is it different from `NOT_STR`, `NOT_STR_OR_LIST_OF_RDD` and `NOT_STR_OR_STRUCT` cases?

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/ml/connect/tuning.py [hunk 2]
```diff
@@ -178,11 +178,12 @@ def _parallelFitTasks(
 
     def get_single_task(index: int, param_map: Any) -> Callable[[], Tuple[int, float]]:
         def single_task() -> Tuple[int, float]:
-            # Active session is thread-local variable, in background thread the active session
-            # is not set, the following line sets it as the main thread active session.
-            active_session._jvm.SparkSession.setActiveSession(  # type: ignore[union-attr]
-                active_session._jsparkSession  # type: ignore[union-attr]
-            )
+            if not is_remote():
```
Comment (HyukjinKwon) [inline python/pyspark/ml/connect/tuning.py:181]: cc @WeichenXu123 FYI the related tests were `pyspark.ml.tests.connect.test_connect_tuning.CrossValidatorTestsOnConnect`.

**Comment by HyukjinKwon** [comment]
cc @cdkrot @grundprinzip too

File: python/pyspark/sql/connect/session.py [hunk 3]
```diff
@@ -236,6 +237,38 @@ def __init__(self, connection: Union[str, ChannelBuilder], userId: Optional[str]
...
+        active :class:`SparkSession` when they are not set yet.
+        """
+        with cls._lock:
+            if cls._default_session is None:
+                cls._default_session = session
+        if getattr(cls._active_session, "session", None) is None:
+            cls._active_session.session = session
+
+    @classmethod
+    def getActiveSession(cls) -> Optional["SparkSession"]:
```
Comment (allisonwang-db) [inline python/pyspark/sql/connect/session.py:253]: Is this also a public API? I found it a bit confusing that this `.getActiveSession()` returns the active session but `.active()` returns the active or the default session (and throws an exception if no active session). Is the API the same as non Spark Connect one?

**Review by allisonwang-db — COMMENTED** [review]
udtf change looks good!

File: python/pyspark/sql/connect/session.py (hunk 3 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/session.py:253]: yup

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: python/pyspark/sql/session.py [hunk 4]
```diff
@@ -805,6 +847,9 @@ def udtf(self) -> "UDTFRegistration":
 
         .. versionadded:: 3.5.0
 
+        .. versionchanged:: 3.4.0
+            Supports Spark Connect.
```
Comment (ueshin) [inline python/pyspark/sql/session.py:851]: This is introduced in `3.5.0`.

File: python/pyspark/sql/connect/session.py [hunk 5]
```diff
@@ -93,14 +94,13 @@
...
-#  If sessions are created with `SparkSession.builder.create`, it stores
-#  The last created session
-_active_spark_session = None
-
-
 class SparkSession:
+    # The active SparkSession for the current thread
+    _active_session = threading.local()
+    # Reference to the root SparkSession
+    _default_session: Optional["SparkSession"] = None
```
Comment (ueshin) [inline python/pyspark/sql/connect/session.py:101]: The type should be `ClassVar[Optional["SparkSession"]]`?

File: python/pyspark/sql/connect/session.py [hunk 6]
```diff
@@ -93,14 +94,13 @@
...
-# `_active_spark_session` stores the active spark connect session created by
-# `SparkSession.builder.getOrCreate`. It is used by ML code.
-#  If sessions are created with `SparkSession.builder.create`, it stores
-#  The last created session
-_active_spark_session = None
-
-
 class SparkSession:
+    # The active SparkSession for the current thread
+    _active_session = threading.local()
```
Comment (ueshin) [inline python/pyspark/sql/connect/session.py:99]: Need the type hint for this?

File: python/pyspark/sql/connect/session.py [hunk 7]
```diff
@@ -93,14 +94,13 @@
...
+    # Reference to the root SparkSession
+    _default_session: Optional["SparkSession"] = None
+    _lock = RLock()
```
Comment (ueshin) [inline python/pyspark/sql/connect/session.py:102]: ditto, `ClassVar[RLock]]`?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/sql/connect/session.py (hunk 6 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/session.py:99]: I am actually not sure how to type this .. it's a private var so should be fine.

Comment (ueshin) [reply python/pyspark/sql/connect/session.py:100]: I guess `ClassVar[threading.local]`?
But ok, should be fine.

File: python/pyspark/errors/error_classes.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/errors/error_classes.py]: They are type related error classes, so we separated them to inform users all the exact possible types for each cases.

On my second thought, we can also separate for this case, nvm.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.

**Pull request closed without merging**